File: src/main/java/com/hanocybous/ecommercesystem/dto/review/ReviewDto.java
package com.hanocybous.ecommercesystem.dto.review;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.time.LocalDateTime;
import java.util.Objects;

public record ReviewDto(
        Long id,
        String title,
        Integer rating,
        Long productId,
        Long userId,
        String comment,
        LocalDateTime reviewDate) {

    public ReviewDto {
        Objects.requireNonNull(id);
        Objects.requireNonNull(title);
        Objects.requireNonNull(rating);
        Objects.requireNonNull(productId);
        Objects.requireNonNull(userId);
        Objects.requireNonNull(comment);
        Objects.requireNonNull(reviewDate);

        if (rating < 1 || rating > 5) {
            throw new IllegalArgumentException("Rating must be between 1 and 5");
        }
        if (title.isBlank()) {
            throw new IllegalArgumentException("Title must not be blank");
        }
        if (comment.isBlank()) {
            throw new IllegalArgumentException("Comment must not be blank");
        }
        if (title.length() > 100) {
            throw new IllegalArgumentException("Title must not be longer than 100 characters");
        }
        if (comment.length() > 1000) {
            throw new IllegalArgumentException("Comment must not be longer than 1000 characters");
        }

    }

    @Contract(pure = true)
    @Override
    public @NotNull String toString() {
        return "ReviewDto{" +
                "id=" + id +
                ", title='" + title + '\'' +
                ", rating=" + rating +
                ", productId=" + productId +
                ", userId=" + userId +
                ", comment='" + comment + '\'' +
                ", reviewDate=" + reviewDate +
                '}';
    }

}

File: src/main/java/com/hanocybous/ecommercesystem/dto/review/ReviewDto.kt
package com.hanocybous.ecommercesystem.dto.review

import java.time.LocalDateTime

data class ReviewDto(
    val id: Long,
    val title: String,
    val rating: Int,
    val productId: Long,
    val userId: Long,
    val comment: String,
    val reviewDate: LocalDateTime
) {

    init {
        require(title.isNotBlank()) { "Title must not be blank" }
        require(title.length <= 100) { "Title must not be longer than 100 characters" }
        require(rating in 1..5) { "Rating must be between 1 and 5" }
        require(comment.isNotBlank()) { "Comment must not be blank" }
        require(comment.length <= 1000) { "Comment must not be longer than 1000 characters" }
    }

    override fun toString(): String {
        return "ReviewDto(id=$id, " +
                "title='$title', " +
                "rating=$rating, " +
                "productId=$productId, " +
                "userId=$userId, " +
                "comment='$comment', " +
                "reviewDate=$reviewDate)"
    }
}
